Support .cortexignore file to exclude files/directories from cortex project zip

wingkwong opened the pull request:
closes #723 

---

checklist:

- [x] run `make test` and `make lint`
- [x] test manually (i.e. build/push all images, restart operator, and re-deploy APIs)

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cortexlabs/cortex?pullRequest=800) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cortexlabs/cortex?pullRequest=800) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cortexlabs/cortex?pullRequest=800) it.</sub>

deliahu commented on pkg/lib/files/files_test.go:
```diff
@@ -239,12 +259,89 @@ func TestListDirRecursive(t *testing.T) {
+
+func TestReadAllIgnorePatterns(t *testing.T) {
+	tmpDir, err := ioutil.TempDir("", "cortexignore-test")
+	if err != nil {
```
`require.NoError(t, err)` (here and below) might be a little cleaner, but not a big deal either way, so feel free to make the change or not based on your preference

wingkwong replied:
Agree. It looks cleaner.

deliahu commented on pkg/lib/files/files_test.go:
```diff
@@ -239,12 +259,89 @@ func TestListDirRecursive(t *testing.T) {
+	if len(di) != 7 {
+		t.Fatalf("Expected 5 entries, got %v", len(di))
+	}
+	if di[0] != "test1" {
```
You can use `require.Equal(t, di[0], "test1")` here and below if you'd like

wingkwong replied:
Agree.

deliahu commented on pkg/lib/zip/zip.go:
```diff
@@ -193,7 +193,8 @@ func addFileToZip(fileInput *FileInput, zipInput *Input, archive *zip.Writer, ad
 
 func addDirToZip(dirInput *DirInput, zipInput *Input, archive *zip.Writer, addedPaths strset.Set) error {
-	paths, err := files.ListDirRecursive(dirInput.Source, true, dirInput.IgnoreFns...)
+	emptyExcludes := make([]string, 0)
```
Will this also work if it's `var emptyExcludes []string` (i.e. if it's `nil`)? It seems to me like it will, just wanted to double check. Also using `nil` will allocate slightly less memory.

wingkwong replied:
Yep. Just tested.

wingkwong commented on pkg/lib/files/files.go:
```diff
@@ -506,7 +508,29 @@ func ListDirRecursive(dir string, relative bool, ignoreFns ...IgnoreFn) ([]strin
+			if len(excludes) > 0 {
+				ignore := false
+				for _, exclude := range excludes {
+					match, _ := filepath.Match(exclude, path)
```
Agree.

wingkwong commented on pkg/lib/files/files.go:
```diff
@@ -558,3 +582,57 @@ func ReadReqFile(r *http.Request, fileName string) ([]byte, error) {
+// ReadAll reads a .cortexignore file and returns the list of file patterns
+// to ignore. Note this will trim whitespace from each line as well
+// as use GO's "clean" func to get the shortest/cleanest path for each.
+func ReadAllIgnorePatterns(reader io.Reader) ([]string, error) {
```
I was following .dockerignore design. Your way seems to be simple tho.

wingkwong commented:
> @wingkwong thank you for your contribution, this is a great feature!
> 
> I made a few comments on the pull request, but nothing major, and most of them are up to you if you want to make the changes I mentioned. I also pushed a few very small changes to your branch.
> 
> I had two questions regarding the behavior of `.cortexignore`:
> 
> 1. As I was running the code, it seemed to me that it didn't work quite as I expected. I was in the `examples/tensorflow/iris-classifier` directory, I created `.cortexignore` with `*.md`, but `README.md` was still included in the project files (I tested this by printing out `projectPaths` in `cli/cmd/deploy.go`).
> 2. I wasn't able to determine by reading the code where a leading `!` is handled (I didn't see mention of it in [filepath.Match](https://golang.org/pkg/path/filepath/#Match)).
> 
> Please let me know your thoughts!

1.  If *.md is in cortexignore.md
https://github.com/wingkwong/cortex/blob/ebfd26301609042f9b3b6c52e2cc3ee00d66813e/pkg/lib/files/files_test.go#L188

and I expect there is a README.md. The test case will be failed 
![image](https://user-images.githubusercontent.com/35857179/74144875-81560800-4c38-11ea-9412-3bb57ac4bdf5.png)

so it should be able to ignore it.

2. the logic comes from 
https://github.com/wingkwong/cortex/blob/b0546616882bc9d17380c551b106a515ca91896e/pkg/lib/files/files.go#L615

deliahu commented:
Thanks for addressing the comments!

For 1, I realized the root of the issue was that were were matching against the relative paths, but when `ListDirRecursive()` was called with `relative` set to `false` (as it is in `cli/cmd/deploy.go`), then it was not working as expected.

For 2, I did see this logic, however it seems like this is just for parsing the patterns, but the `!` was not being interpreted by the matcher.

It seemed like even according to [Docker's documentation](https://docs.docker.com/engine/reference/builder/#dockerignore-file), they had to do some workarounds on top of Go's `filepath.Match()`. So I searched for a replacement to the matcher and found this one: https://godoc.org/github.com/denormal/go-gitignore. I switch to it in to your PR, and also tried using the `files.IgnoreFn()` pattern, what do you think of it?

wingkwong commented:
> Thanks for addressing the comments!
> 
> For 1, I realized the root of the issue was that were were matching against the relative paths, but when `ListDirRecursive()` was called with `relative` set to `false` (as it is in `cli/cmd/deploy.go`), then it was not working as expected.
> 
> For 2, I did see this logic, however it seems like this is just for parsing the patterns, but the `!` was not being interpreted by the matcher.
> 
> It seemed like even according to [Docker's documentation](https://docs.docker.com/engine/reference/builder/#dockerignore-file), they had to do some workarounds on top of Go's `filepath.Match()`. So I searched for a replacement to the matcher and found this one: https://godoc.org/github.com/denormal/go-gitignore. I switch to it in to your PR, and also tried using the `files.IgnoreFn()` pattern, what do you think of it?

I like this approach. It looks more cleaner!

deliahu commented:
Great! I'll send it to @vishalbollu for review, and then we'll merge it in

vishalbollu commented on pkg/lib/files/files.go:
```diff
@@ -347,6 +348,23 @@ func IgnoreSpecificFiles(absPaths ...string) IgnoreFn {
 	}
 }
 
+func GitIgnoreFn(gitIgnorePath string) (IgnoreFn, error) {
```
I think the .cortexignore file should be excluded from the zip because I don't think the APIs should be redeployed if the .cortexignore file changes without any of the project files changing.

deliahu replied:
`.cortexignore` will be excluded from the zip since we also use `files.IgnoreHiddenFiles` (so all files starting with . will be ignored)

deliahu commented on cli/cmd/deploy.go:
```diff
@@ -106,6 +107,15 @@ func deploy(configPath string, force bool) {
+		if err != nil {
+			exit.Error(err)
+		}
+		ignoreFns = append(ignoreFns, cortexIgnore)
```
Great point! I'll check now, I think we will need to reverse the order

deliahu commented:
@wingkwong I will merge it now, thank you for your contribution!
